### ISSUE TITLE
BUG due to matrixStats (>= 0.50.0): Add named arguments in matrixStats calls

### ISSUE THREAD

**Issue by HenrikBengtsson** [ccagc/QDNAseq#16]
In [matrixStats (>= 0.50.0)](https://cran.r-project.org/package=matrixStats), which is now on CRAN, subsetted arguments `idxs`, `rows` and `cols` where introduced to most functions.  These has been inserted just after the arguments that are to be subsetted, which typically are the first one (e.g. `x`) or first two arguments (e.g. `x` and `w`).  I have detected code in your package that is likely to give an error or the wrong answer because it calls matrixStats functions assuming a certain order of arguments.

More specifically, I believe following functions needs to be updated _as soon as possible_ by using named arguments as proposed:
```r
$QDNAseq$.binReadCountsPerSample
NOW: counts <- binCounts(hits[[chromosome]], chromosomeBreaks)
SUGGESTION: counts <- binCounts(hits[[chromosome]], bx=chromosomeBreaks)
```

**Comment by daoud-sie** [comment]
Thanks for the suggestion. It is fixed. Closing the issue.

**Comment by ilarischeinin** [comment]
How do the branches work nowadays? This fix looks to be in Bioconductor's svn, but it's not here in the `master` branch. Is `release` the Bioconductor release version and `develop` the Bioconductor development version? What is `master` then? It seems like it's just left behind.

**Comment by daoud-sie** [comment]
http://nvie.com/posts/a-successful-git-branching-model/

Following this model the master branch indeed seems a bit left behind. With the release of the next version all the changes accumulated in the develop branch will be merged with the master branch.

The release branch needs to be in sync with the bioconductor release, which is not the case yet. Will need to fix that first before I can continue.


-------- Original message --------

**Comment by ilarischeinin** [comment]
Ah!

I guess we just have quite different interpretations on how to combine that model and Bioconductor's release model.

> We consider origin/master to be the main branch where the source code of HEAD always reflects a production-ready state.

My view is that `master` corresponds to the Bioconductor development version, and `develop` is the integration branch where changes from different developers are integrated before being released to the public (to the development Bioconductor). The release version of Bioconductor I consider something that's a separate entity that is created mainly for posterity and reproducibility. (i.e. `develop`= pre-Bioconductor, `master` = Bioconductor development version, `release`/`release-x.y` = Bioconductor release version)

I guess your take is that what is "production-ready", is the release, not development version of Bioconductor. Therefore branch `develop` is what corresponds to the development version of Bioconductor. However, I don't think the repo's current state actually represents that model. If `develop` = Bioconductor development version, `master` should be the Bioconductor release version. `release`/`release-x.y` branches would therefore not be needed, or they could be created as pure archival branches.

In any case, at the moment `master` is behind both development and release versions of Bioconductor. Regardless of which one considers to be the "production-ready" one, I don't think that should ever happen according to that branching model.

**Comment by daoud-sie** [comment]
Ok, so what you’re saying is that we can follow the git workflow I referenced but:

Origin/master should be regarded as production-ready/development.

This wasn’t clear to me by just looking at the first figure. I will merge back stable states of develop to master.

The link with BioC is as follows:


1.       Origin/master ->  BioC development

2.       Origin/release/release.x.y -> BioC release.x.y

**Comment by ilarischeinin** [comment]
> The link with BioC is as follows:
> 1.       Origin/master ->  BioC development

Um, but it's not. This matrixStats issue is fixed in [Bioconductor development](https://hedgehog.fhcrc.org/bioconductor/trunk/madman/Rpacks/QDNAseq/R/binReadCounts.R):

> counts <- binCounts(hits[[seqNameI]], bx=chromosomeBreaks)
> ...
> counts <- binCounts(hits[[chromosome]], bx=chromosomeBreaks)

But it's not fixed in [origin/master](https://github.com/ccagc/QDNAseq/blob/master/R/binReadCounts.R):

> counts <- binCounts(hits[[seqNameI]], chromosomeBreaks)
> ...
> counts <- binCounts(hits[[chromosome]], chromosomeBreaks)

At the moment, origin/master is behind "production-ready", so it's not following that linked branching model.

**Comment by ilarischeinin** [comment]
I mean no matter how you define "production-ready" (Bioconductor development or release), origin/master is behind, because the fix is both in Bioconductor development (above) and [Bioconductor release](https://hedgehog.fhcrc.org/bioconductor/branches/RELEASE_3_2/madman/Rpacks/QDNAseq/R/binReadCounts.R):

> counts <- binCounts(hits[[chromosome]], bx=chromosomeBreaks)

**Comment by HenrikBengtsson** [comment]
Regardless of what Bioconductor connection you are using, I found the [git flow](http://danielkummer.github.io/git-flow-cheatsheet/) model extremely useful. The rule of thumb is that:

* The `master` branch should reflect what is "released" to user.
* The `develop` branch should reflect what is under active development (and not released).
* Never commit to `master`!  Only commit to `develop` and make sure things are as expected. Only the merge into `master`, i.e. `git checkout master; git merge develop; git push; git checkout develop`.  (the last command lower the risk of committing to `master` by mistake; only have it checkout out when merging).

I'm pretty sure I mentioned this to you in the past and you did set that up.  *Note difference between Bioc 'devel' repos and Git `develop` branch*!  Then I would recommend that you connect `master` to the Bioconductor 'devel' repository.  When you used the Bioc git-svn bridge, I'm pretty sure this is what you had/used at the time.

What makes it a bit messy is that the Bioc "release" repository is now also going to be connected to the same Git repository.  For Bioc 'release' repositories, I would probably use a `Bioc-release` branch and make sure that syncs with Bioc "release".  If you find it more clear, you could use a `Bioc-devel` branch` and make sure that syncs with Bioc "devel".  Then I would simply keep the `master` and `Bioc-devel` in perfect sync.  Same rule applies as before, never commit to `master` nor `Bioc-devel` - only merge.   The last piece is how to update `Bioc-release`?  Since it should happen very rarely, you could take the risk of allowing yourself to edit that branch directly, but ideally use follow the "git flow" rule of never commit to production and instead create a separate branch whenever you need to update that.  Git flow support the model of "hot fixes", e.g.
```
git checkout Bioc-release
git checkout -b hotfix/typos
<edit the files and check with R CMD check>
git commit -a -m "Fixed typos"
git checkout Bioc-release
git merge hotfix/typos
git svn commit # ... or whatever you do to commit to Bioc release repos
git branch -D hotfix/typos
```

It looks a mouth full, but point it is really easy to remember if you stick to the git flow model, which has made my life so much easier.  Much less trial and error and no more flavor-of-the-day work processes.

**Comment by daoud-sie** [comment]
Ok. So now the changes are made in "develop", merged with "master" and commited with svn to bioc-devel. What is released in the development version is now reflected in "master", so that's all ok, right?

I also fixed bioc-release-3.2 directly in this case. I wanted to push these changes to github, but was not sure how that would work. The history of the svn release branch seems truncated, making it unclear where this release branched off from the master in github. So I was a bit unsure what to do. 

I now branched a release 3.1 and 3.2, adding the changes made to those release versions. 

It should be more clean now, let me know if you are still seeing issues...
